keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,116 @@
+
+## Motivation
+
+SideCar containers have always been used in some ways but just not formally identified as such, with the introduction of Jobs and Cronjobs it became clear that they're not handled very well. If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish.
+
+The only way around this problem is to manage the sidecar container's lifecycle manually and arrange for it to exit when the main container exits. This is typically achieved by building an ad-hoc signalling mechanism to communicate completion status between containers. Common implementations use a shared scratch volume mounted into all pods, where lifecycle status can be communicated by creating and watching for the presence of files. This pattern has several disadvantages:
+
+* Repetitive lifecycle logic must be rewritten in each instance a sidecar is deployed.
+* Third-party containers typically require a wrapper to add this behaviour, normally provided via an entrypoint wrapper script implemented in the k8s container spec. This adds undesirable overhead and introduces repetition between the k8s and upstream container image specs.
+* The wrapping typically requires the presence of a shell in the container image, so this pattern does not work for minimal containers which ship without a toolchain.
+
+### Goals
+
+Solve issue [25908](https://github.com/kubernetes/kubernetes/issues/25908) so that sidecars can be easily handled without needing modifications to the container itself
+
+### Non-Goals
+
+Changing the way jobs are handled
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+These sidecar containers will be handled the same way as a non-sidecar container in almost all respects apart from the fact that when all non-sidecar containers have terminated all running sidecar containers will also be terminated with a reason of `Completed`.
+
+For example, in a job you can have as many sidecars as you like and all the running sidecars will be terminated when all the non-sidecar containers have terminated, causing your job to finish.
+If all the non-sidecars `Completed` the pod status will be `Succeeded`, if any of the non-sidecars terminated with `Error` the pod status will be `Failed` and if any of the sidecars terminated with `Error` before the non-sidecars terminated, the pod status will be `Failed`.
```

Comment:
This means that a crashy non-critical sidecar could cause the entire pod to be marked Failed, which is not ideal. Perhaps the sidecars could be safely restarted as long as a non-sidecar container is running?

Reply:
That's debatable, depending on the purpose of a sidecar it failing could be quite important and you would want to see that reflected in the pod status, also that's no different from what currently happens. 
The sidecars would still follow the same restart policy as the rest of the pod so for example you can have `restartPolicy: OnFailure` if you desire, but I think making them ignore the restart policy of the pod and always restart is perhaps too presumptive, there may be people who don't want their sidecar to do that.